components/brave_ads/browser/ads_notification_handler.cc

```diff
@@ -120,4 +157,47 @@ const void* AdsNotificationHandler::UserDataKey() {
   return &kUserDataKey;
 }
 
+void AdsNotificationHandler::OperationCompleted(
+    const std::string& notification_id) {
+
+  auto iter = pending_callbacks_.find(notification_id);
+  std::move(iter->second).Run();
+  pending_callbacks_.erase(iter);
+#if defined(OS_ANDROID)
+  StartShutDownTimerIfNecessary(notification_id);
+#endif
+}
+
+#if defined(OS_ANDROID)
+bool AdsNotificationHandler::IsHeadless(){
+  auto state = base::android::ApplicationStatusListener::GetState();
+  bool uiless = (state ==
+      base::android::ApplicationState::APPLICATION_STATE_UNKNOWN ||
+      state ==
+      base::android::ApplicationState::APPLICATION_STATE_HAS_DESTROYED_ACTIVITIES);
+  return uiless;
+}
+
+void AdsNotificationHandler::StartShutDownTimerIfNecessary (
+    const std::string & last_processed_notification_id){
+  if (IsHeadless() && last_dismissed_notification_id_ ==
+      last_processed_notification_id){
+    last_dismissed_notification_id_ = "";
+    //wait and close the browser if running headless
+    no_ui_shutdown_timer_.Stop();
```

Comment:
I don't think you need to explicitly call `Stop`.

Reply:
Removed